utils/fs/memory/memory.go

```diff
@@ -0,0 +1,333 @@
+}
+
+func (fs *Memory) Open(filename string) (fs.File, error) {
+	return fs.OpenFile(filename, os.O_RDONLY, 0)
+}
+
+func (fs *Memory) OpenFile(filename string, flag int, perm os.FileMode) (fs.File, error) {
+	fullpath := fs.Join(fs.base, filename)
+	f, ok := fs.s.files[fullpath]
+	if !ok && flag&os.O_CREATE == 0 {
+		return nil, os.ErrNotExist
+	}
+
+	if f == nil {
+		fs.s.files[fullpath] = newFile(fs.base, fullpath, flag)
+		return fs.s.files[fullpath], nil
+	}
+
+	n := newFile(fs.base, fullpath, flag)
+	n.c = f.c
+
+	if flag&os.O_APPEND != 0 {
+		n.p = n.c.size
+	}
+
+	if flag&os.O_TRUNC != 0 {
+		n.c.Truncate()
+	}
+
+	return n, nil
+}
+
+func (fs *Memory) Stat(filename string) (fs.FileInfo, error) {
+	fullpath := fs.Join(fs.base, filename)
+
+	if _, ok := fs.s.files[filename]; ok {
+		return newFileInfo(fs.base, fullpath, fs.s.files[filename].c.size), nil
+	}
+
+	info, err := fs.ReadDir(filename)
+	if err == nil && len(info) != 0 {
+		return newFileInfo(fs.base, fullpath, int64(len(info))), nil
+	}
+
+	return nil, os.ErrNotExist
+}
+
+func (fs *Memory) ReadDir(base string) (entries []fs.FileInfo, err error) {
+	base = fs.Join(fs.base, base)
+
+	dirs := make(map[string]bool, 0)
+	for fullpath, f := range fs.s.files {
+		if !strings.HasPrefix(fullpath, base) {
+			continue
+		}
+
+		fullpath, _ = filepath.Rel(base, fullpath)
+		parts := strings.Split(fullpath, string(separator))
+
+		if len(parts) != 1 {
+			dirs[parts[0]] = true
+			continue
+		}
+
+		entries = append(entries, newFileInfo(fs.base, fullpath, f.c.size))
+	}
+
+	for path := range dirs {
+		entries = append(entries, &fileInfo{
+			name:  path,
+			isDir: true,
+		})
+	}
+
+	return
+}
+
+func (fs *Memory) TempFile(dir, prefix string) (fs.File, error) {
+	filename := fmt.Sprintf("%s_%d", prefix, time.Now().UnixNano())
+	fullpath := fs.Join(fs.base, dir, filename)
+	return fs.Create(fullpath)
+}
+
+func (fs *Memory) Rename(from, to string) error {
+	from = fs.Join(fs.base, from)
+	to = fs.Join(fs.base, to)
+
+	if _, ok := fs.s.files[from]; !ok {
+		return os.ErrNotExist
+	}
+
+	fs.s.files[to] = fs.s.files[from]
+	fs.s.files[to].BaseFilename = to
+	delete(fs.s.files, from)
+
+	return nil
+}
+
+func (fs *Memory) Remove(filename string) error {
+	if _, err := fs.Stat(filename); err != nil {
+		return err
+	}
+
+	fullpath := fs.Join(fs.base, filename)
+	delete(fs.s.files, fullpath)
+	return nil
+}
+
+func (fs *Memory) clean(path string) string {
+	if len(path) <= 1 {
+		if path != string(separator) {
+			return path
+		}
+
+		return ""
+	}
+
+	if path[0] == separator {
+		path = path[1:]
+	}
+
+	l := len(path)
+	if path[l-1] == separator {
+		path = path[:l-1]
+	}
+
+	return path
+}
+
+func (fs *Memory) Join(elem ...string) string {
+	return filepath.Join(elem...)
+}
+
+func (fs *Memory) Dir(path string) fs.Filesystem {
+	return &Memory{
+		base: fs.Join(fs.base, path),
+		s:    fs.s,
+	}
+}
+
+func (fs *Memory) Base() string {
+	return fs.base
+}
+
+type file struct {
```

Comment:
errr, I don't like, is everything is part of the same thing, and since you don't have test only for the files, is weird have a `file.go` without the `file_test.go`